app/src/org/odk/collect/android/activities/FormEntryActivity.java

```diff
@@ -2085,14 +2100,16 @@ public void onClick(DialogInterface dialog, int i) {
      *
      * @param exit if set, will exit program after save.
      * @param complete has the user marked the instances as complete?
-     * @param updatedSaveName set name of the instance's content provider, if non-null
+     * @param updatedSaveName set name of the instance's content provider, if
+     *                        non-null
      * @param headless is this running as a GUI-less service
      *
      * @return Did the data save successfully?
      */
-    private boolean saveDataToDisk(boolean exit, boolean complete, String updatedSaveName, boolean headless) {
+    private boolean saveDataToDisk(boolean exit, boolean complete,
+                                   String updatedSaveName, boolean headless) {
         // save current answer
-        if (!saveAnswersForCurrentScreen(EVALUATE_CONSTRAINTS, complete)) {
+        if (!saveAnswersForCurrentScreen(EVALUATE_CONSTRAINTS, complete, headless)) {
```

Comment:
Session-expiration form saving will abort if the current screen has a value that violates the constraint. We could try to roll back that violating value and save, should we attempt this?